### ISSUE TITLE
Missing "mount" directory on provisioned persistent volume

### ISSUE THREAD

**Issue by 2martens** [hetznercloud/csi-driver#534]
I installed a MicroK8s cluster on a Hetzner server and integrated the CSI driver helm chart. The PV creation itself works and it is mounted to the server by Hetzner. But the directory contains only `vol_data.json` and not a subdirectory `mount` which is used by Kubernetes as mount point for persistent volumes.

I had to manually create this directory via SSH access on the host and then the pod was able to actually mount the PV into itself. The PV was already bound to the PVC. 

PVC created following a volume claim template from Prometheus Operator:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: >-
    prometheus-prometheus-kube-prometheus-prometheus-db-prometheus-prometheus-kube-prometheus-prometheus-0
  namespace: observability
  uid: dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767
  resourceVersion: '279650'
  creationTimestamp: '2023-11-06T21:56:13Z'
  labels:
    app.kubernetes.io/instance: prometheus-kube-prometheus-prometheus
    app.kubernetes.io/managed-by: prometheus-operator
    app.kubernetes.io/name: prometheus
    operator.prometheus.io/name: prometheus-kube-prometheus-prometheus
    operator.prometheus.io/shard: '0'
    prometheus: prometheus-kube-prometheus-prometheus
  annotations:
    pv.kubernetes.io/bind-completed: 'yes'
    pv.kubernetes.io/bound-by-controller: 'yes'
    volume.beta.kubernetes.io/storage-provisioner: csi.hetzner.cloud
    volume.kubernetes.io/selected-node: k8s-test-node-0
    volume.kubernetes.io/storage-provisioner: csi.hetzner.cloud
  finalizers:
    - kubernetes.io/pvc-protection
  managedFields:
    - manager: kubelite
      operation: Update
      apiVersion: v1
      time: '2023-11-06T21:56:18Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:pv.kubernetes.io/bind-completed: {}
            f:pv.kubernetes.io/bound-by-controller: {}
            f:volume.beta.kubernetes.io/storage-provisioner: {}
            f:volume.kubernetes.io/selected-node: {}
            f:volume.kubernetes.io/storage-provisioner: {}
          f:labels:
            .: {}
            f:app.kubernetes.io/instance: {}
            f:app.kubernetes.io/managed-by: {}
            f:app.kubernetes.io/name: {}
            f:operator.prometheus.io/name: {}
            f:operator.prometheus.io/shard: {}
            f:prometheus: {}
        f:spec:
          f:accessModes: {}
          f:resources:
            f:requests:
              .: {}
              f:storage: {}
          f:storageClassName: {}
          f:volumeMode: {}
          f:volumeName: {}
    - manager: kubelite
      operation: Update
      apiVersion: v1
      time: '2023-11-06T21:56:18Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:accessModes: {}
          f:capacity:
            .: {}
            f:storage: {}
          f:phase: {}
      subresource: status
  selfLink: >-
    /api/v1/namespaces/observability/persistentvolumeclaims/prometheus-prometheus-kube-prometheus-prometheus-db-prometheus-prometheus-kube-prometheus-prometheus-0
status:
  phase: Bound
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 10Gi
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  volumeName: pvc-dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767
  storageClassName: hcloud-volumes
  volumeMode: Filesystem
```

The persistent volume bound to it:
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pvc-dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767
  uid: bfc9c55c-5336-475b-8c37-f222a42c1a70
  resourceVersion: '279655'
  creationTimestamp: '2023-11-06T21:56:18Z'
  annotations:
    pv.kubernetes.io/provisioned-by: csi.hetzner.cloud
    volume.kubernetes.io/provisioner-deletion-secret-name: ''
    volume.kubernetes.io/provisioner-deletion-secret-namespace: ''
  finalizers:
    - kubernetes.io/pv-protection
    - external-attacher/csi-hetzner-cloud
  managedFields:
    - manager: csi-provisioner
      operation: Update
      apiVersion: v1
      time: '2023-11-06T21:56:18Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:pv.kubernetes.io/provisioned-by: {}
            f:volume.kubernetes.io/provisioner-deletion-secret-name: {}
            f:volume.kubernetes.io/provisioner-deletion-secret-namespace: {}
        f:spec:
          f:accessModes: {}
          f:capacity:
            .: {}
            f:storage: {}
          f:claimRef:
            .: {}
            f:apiVersion: {}
            f:kind: {}
            f:name: {}
            f:namespace: {}
            f:resourceVersion: {}
            f:uid: {}
          f:csi:
            .: {}
            f:driver: {}
            f:fsType: {}
            f:volumeAttributes:
              .: {}
              f:storage.kubernetes.io/csiProvisionerIdentity: {}
            f:volumeHandle: {}
          f:nodeAffinity:
            .: {}
            f:required: {}
          f:persistentVolumeReclaimPolicy: {}
          f:storageClassName: {}
          f:volumeMode: {}
    - manager: kubelite
      operation: Update
      apiVersion: v1
      time: '2023-11-06T21:56:18Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:phase: {}
      subresource: status
    - manager: csi-attacher
      operation: Update
      apiVersion: v1
      time: '2023-11-06T21:56:19Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:finalizers:
            v:"external-attacher/csi-hetzner-cloud": {}
  selfLink: /api/v1/persistentvolumes/pvc-dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767
status:
  phase: Bound
spec:
  capacity:
    storage: 10Gi
  csi:
    driver: csi.hetzner.cloud
    volumeHandle: '100059611'
    fsType: ext4
    volumeAttributes:
      storage.kubernetes.io/csiProvisionerIdentity: 1699301579169-8081-csi.hetzner.cloud
  accessModes:
    - ReadWriteOnce
  claimRef:
    kind: PersistentVolumeClaim
    namespace: observability
    name: >-
      prometheus-prometheus-kube-prometheus-prometheus-db-prometheus-prometheus-kube-prometheus-prometheus-0
    uid: dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767
    apiVersion: v1
    resourceVersion: '279632'
  persistentVolumeReclaimPolicy: Retain
  storageClassName: hcloud-volumes
  volumeMode: Filesystem
  nodeAffinity:
    required:
      nodeSelectorTerms:
        - matchExpressions:
            - key: csi.hetzner.cloud/location
              operator: In
              values:
                - fsn1
```

**Comment by apricote** [comment]
Mh, this should work and we test this in k3s in our CI.

- Could you check the logs of the Kubelet & the `hcloud-csi-node` Pod on the Server during the time this issue happened?
- Do you have a `VolumeAttachment` resource?
- Did you confirm that your volume is successfully mounted at the right location now? If you only created the `mount` folder, it might not be properly mounted. You can check by running `mount | grep /var/lib/kubelet`

**Comment by 2martens** [comment]
Logs from the driver container when publishing a volume:
```
level=info ts=2023-11-07T09:04:06.390511082Z component=linux-mount-service msg="formatting disk" disk=/dev/disk/by-id/scsi-0HC_Volume_100061990 fstype=ext4
level=info ts=2023-11-07T09:04:06.483923454Z component=linux-mount-service msg="publishing volume" target-path=/var/snap/microk8s/common/var/lib/kubelet/pods/221c44e1-7ede-4aca-9683-09f1a6475f48/volumes/kubernetes.io~csi/pvc-a6645c33-7bf0-47e4-bc92-11e6390dcf99/mount device-path=/dev/disk/by-id/scsi-0HC_Volume_100061990 fs-type=ext4 block-volume=false readonly=false mount-options= encrypted=false
```

When trying to unpublish a volume:
```
level=info ts=2023-11-07T08:59:40.902578284Z component=linux-mount-service msg="unpublishing volume" target-path=/var/snap/microk8s/common/var/lib/kubelet/pods/11f7edea-571c-4f4f-8368-2bae20e59e63/volumes/kubernetes.io~csi/pvc-dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767/mount device-path=
W1107 08:59:40.902628       1 mount_helper_common.go:34] Warning: mount cleanup skipped because path does not exist: /var/snap/microk8s/common/var/lib/kubelet/pods/11f7edea-571c-4f4f-8368-2bae20e59e63/volumes/kubernetes.io~csi/pvc-dbb1b4b6-0df5-45eb-90e6-ca6d4ae3e767/mount
```

Volume Attachments:
```
NAME                                                                   ATTACHER            PV                                         NODE              ATTACHED   AGE
csi-f75a4701ddfcbfc39356c54ffe92a914f2ad0dc2c7f36011c164d9f30fffc2c0   csi.hetzner.cloud   pvc-a6645c33-7bf0-47e4-bc92-11e6390dcf99   k8s-test-node-0   true       22h
```
If I look up the mounts, I see this:
```
/dev/sda1 on /var/snap/microk8s/common/var/lib/kubelet/pods/d9ff7160-10e2-4099-a84b-64fff56abe9a/volume-subpaths/pvc-a6645c33-7bf0-47e4-bc92-11e6390dcf99/prometheus/2 type ext4 (rw,relatime)
/dev/sda1 on /var/snap/microk8s/common/var/lib/kubelet/pods/d9ff7160-10e2-4099-a84b-64fff56abe9a/volume-subpaths/pvc-a6645c33-7bf0-47e4-bc92-11e6390dcf99/thanos-sidecar/0 type ext4 (rw,relatime)
```

Under the path `/var/snap/microk8s/common/var/lib/kubelet/pods/d9ff7160-10e2-4099-a84b-64fff56abe9a/volumes/kubernetes.io~csi/pvc-a6645c33-7bf0-47e4-bc92-11e6390dcf99`, I see now `vol_data.json` and my created directory `mount` which contains `prometheus-db` which was created by the pod. The volume-subpaths seen in the mounts point to that directory for both containers.

But the `mount` directory misses each time a new volume is created by the StorageClass and then bound to the PVC.

**Comment by apricote** [comment]
### sda is root volume

`/dev/sda` should be the root volume of your server, mounted at `/` in the default configuration. I would expect these are `emptyDir` or other emphemeral volumes.

### Identifying Volumes

You can check which device parts belong to your volumes through `lsblk -o path,vendor,serial`. The vendor is `HC` for Cloud Volumes, and the serial is the Volume ID as returned by the API.

 ###  Successful mount

In our dev environment using `k3s` the output of `lsblk` & `mount` looks like this for a successfully mounted PVC using the csi-driver:

```
root@csi-dev-1:~# lsblk -o path,vendor,serial
PATH       VENDOR   SERIAL
/dev/sda   QEMU     38953182
/dev/sda1           
/dev/sda14          
/dev/sda15          
/dev/sdb   HC       100065425
/dev/sr0   QEMU     QM00001
root@csi-dev-1:~# hcloud v^C
root@csi-dev-1:~# mount | grep sdb
/dev/sdb on /var/lib/kubelet/pods/27bd891d-2df6-4b55-aa6e-55e6acc5efd8/volumes/kubernetes.io~csi/pvc-863d5cbc-999a-49ee-b402-1ca1d0d85b98/mount type ext4 (rw,relatime)
```

**I can confirm that I do not see an actual mounted volume on a test microk8s installation.**

**Comment by apricote** [comment]
Okay, the issue is that microk8s uses a non-standard path for the Kubelet. I was able to get a successful mount by following this comment: https://github.com/hetznercloud/csi-driver/issues/213#issuecomment-900650971

#516 added support for configuring the kubelet path to the Helm Chart. I have opened another PR to add the required value for microk8s to the docs (#536) and will publish a new release once that is merged.

**Comment by apricote** [comment]
Works now for me with the following command. Could you try to upgrade and set the required value @2martens?

```
microk8s helm upgrade -install hcloud-csi hcloud/hcloud-csi -n kube-system --version 2.6.0 --set node.kubeletDir=/var/snap/microk8s/common/var/lib/kubelet
```

**Comment by 2martens** [comment]
I upgraded and set the value. The problem no longer appears.
